src/ucp/core/ucp_context.c

```diff
@@ -428,17 +431,127 @@ ucs_status_t ucp_config_read(const char *env_prefix, const char *filename,
 }
 
+static void ucp_cached_key_release(ucs_config_cached_key_t *key_val)
+{
+    if (key_val == NULL) {
+        return;
+    }
+
+    free(key_val->key);
+    free(key_val->val);
+    free(key_val->subfield_prefix);
+    ucs_free(key_val);
+}
+
+static ucs_status_t
+ucp_cached_key_alloc(ucs_config_cached_key_t **key_val,
+                     const char *key, const char *val,
+                     const char *subfield_prefix,
+                     size_t match_level)
+{
+    ucs_config_cached_key_t *kv;
+
+    kv = ucs_calloc(1, sizeof(*kv), "cached config key");
+    if (kv == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    kv->match_level = match_level;
+    kv->key         = strdup(key);
+    kv->val         = strdup(val);
+
+    if (kv->key == NULL || kv->val == NULL) {
+        goto err_free_key;
+    }
+
+    if (match_level == 0) {
+        kv->subfield_prefix = strdup(subfield_prefix);
+        if (kv->subfield_prefix == NULL) {
+            goto err_free_key;
+        }
+    }
+
+    *key_val = kv;
+    return UCS_OK;
+
+err_free_key:
+    ucp_cached_key_release(kv);
+    return UCS_ERR_NO_MEMORY;
+}
+
 void ucp_config_release(ucp_config_t *config)
 {
+    ucs_config_cached_key_t *key_val;
+
+    while (!ucs_list_is_empty(&config->list)) {
+        key_val = ucs_list_extract_head(&config->list, typeof(*key_val), list);
+        ucp_cached_key_release(key_val);
+    }
+
     ucs_config_parser_release_opts(config, ucp_config_table);
     ucs_free(config->env_prefix);
     ucs_free(config);
 }
 
+static UCS_LIST_HEAD(key_list);
+
+static void UCS_F_DTOR ucp_config_free_keys(void)
+{
+    ucs_config_parser_release_keys(&key_list);
+}
+
+static ucs_status_t ucp_config_setup_keys(void)
+{
+    static ucs_init_once_t setup_key_list = UCS_INIT_ONCE_INITIALIZER;
+    static ucs_status_t ucs_status;
+    uct_component_h *components;
+    unsigned num_components;
+
+    UCS_INIT_ONCE(&setup_key_list) {
+        ucs_status = uct_query_components(&components, &num_components);
+        if (ucs_status == UCS_OK) {
+            uct_release_component_list(components);
+        }
+
+        ucs_status = ucs_config_setup_keys(&key_list);
+    }
+    return ucs_status;
+}
+
 ucs_status_t ucp_config_modify(ucp_config_t *config, const char *name,
                                const char *value)
 {
-    return ucs_config_parser_set_value(config, ucp_config_table, name, value);
+    ucs_status_t ucs_status;
+    size_t match_level;
+    char subfield_prefix[64];
+    ucs_config_cached_key_t *key_val;
+
+    ucs_status = ucs_config_parser_set_value(config, ucp_config_table, name,
+                                             value);
+    if (ucs_status == UCS_OK) {
+        return UCS_OK;
+    }
+
+    ucs_status = ucp_config_setup_keys();
+    if (ucs_status != UCS_OK) {
+        return UCS_ERR_CANCELED;
+    }
+
+    memset(subfield_prefix, 0, sizeof(subfield_prefix));
+    ucs_status = ucs_config_parser_audit_key(&key_list, name, &match_level,
```

Comment:
There are a lot of code to check the key, this makes the code much complex. Maybe we don't need to check the key here since in uct_config_modify() the key will also be checked. Then we can remove a lot of related code. @yosefe WDYT?